File: 218-the-skyline-problem/218-the-skyline-problem.java
class Solution {
    public List<List<Integer>> getSkyline(int[][] bs) {
        List<int[]> lines = new ArrayList<>();
        for(int[] b : bs) {
            lines.add(new int[]{b[0], -b[2]});
            lines.add(new int[]{b[1], b[2]});
        }
        Collections.sort(lines, (a, b) -> a[0] == b[0] ? a[1] - b[1] : a[0] - b[0]);
        PriorityQueue<Integer> pq = new PriorityQueue<>((a, b) -> b - a);
        pq.offer(0);
        int preMax = 0;
        List<List<Integer>> res= new ArrayList<>();
        for(int[] l : lines) {
            if(l[1] < 0) {
                pq.add(-l[1]);
            } else {
                pq.remove(l[1]);
            }
            int currMax = pq.peek();
            if(currMax != preMax) {
                res.add(List.of(l[0], currMax));
                preMax = currMax;
            }
        }
        return res;
    }
}

File: 218-the-skyline-problem/218-the-skyline-problem.py
class Solution:
    def getSkyline(self, buildings: List[List[int]]) -> List[List[int]]:
        events = [(L, -H, R) for L, R, H in buildings]
        events += [(R, 0, 0) for _, R, _ in buildings]
        events.sort()
        
        res= [[0, 0]]
        live = [(0, float('inf'))]
        
        for pos, negh, r in events:
            while live[0][1] <= pos:
                heapq.heappop(live)
            if negh: 
                heapq.heappush(live, (negh, r))
            if res[-1][1] != -live[0][0]:
                res += [[pos, -live[0][0]]]
        return res[1:]
            